### PR TITLE
[MLIR] Fix hlo to lhlo conversion amid std.constant ops

### LINKED ISSUE #39895: [MLIR] xla hlo -> lhlo conversion issue when operand is a constant tensor
With master branch as of 831a55584749593400807e0baa7478476b5dbc70 (May 26):
The xla hlo to lhlo lowering doesn't convert completely when the operand of an op (in this example below, that of broadcast_in_dim) is a constant tensor. To reproduce, please use:

```
func @main(%arg0: tensor<4x64x128x3xf32>, %arg1: tensor<5x3x3x10xf32>) {
  %cst = constant  dense<0.000000e+00> : tensor<f32>
  %0 = "xla_hlo.broadcast_in_dim"(%cst) {broadcast_dimensions = dense<[]> : tensor<0xi64>, name = "broadcast.6"} : (tensor<f32>) -> tensor<4x30x42x10xf32>
  return
}
```

and

```
$ tf-opt -hlo-legalize-to-lhlo constant-tensor-lowering.mlir 
constant-tensor-lowering.mlir:3:8: error: 'xla_lhlo.broadcast_in_dim' op operand #0 must be memref of floating-point or signless integer or complex-type values, but got 'tensor<f32>'
  %0 = "xla_hlo.broadcast_in_dim"(%cst) {broadcast_dimensions = dense<[]> : tensor<0xi64>, name = "broadcast.6"} : (tensor<f32>) -> tensor<4x30x42x10xf32>
       ^
constant-tensor-lowering.mlir:3:8: note: see current operation: "xla_lhlo.broadcast_in_dim"(%cst, %0) {broadcast_dimensions = dense<[]> : tensor<0xi64>, name = "broadcast.6"} : (tensor<f32>, memref<4x30x42x10xf32>) -> ()
```
Using `-print-ir-after-all` reveals the operand for the lhlo broadcast_in_dim op wasn't converted to a memref:

```
tuple.mlir:4:10: error: 'xla_lhlo.broadcast_in_dim' op operand #0 must be memref of floating-point or signless integer or complex-type values, but got 'tensor<f32>'
    %0 = "xla_hlo.broadcast_in_dim"(%cst) {broadcast_dimensions = dense<[]> : tensor<0xi64>, name = "broadcast.6"} : (tensor<f32>) -> tensor<4x30x42x10xf32>
         ^
tuple.mlir:4:10: note: see current operation: "xla_lhlo.broadcast_in_dim"(%cst, %0) {broadcast_dimensions = dense<[]> : tensor<0xi64>, name = "broadcast.6"} : (tensor<f32>, memref<4x30x42x10xf32>) -> ()
// *** IR Dump After mlir::detail::VerifierPass Failed ***


"module"() ( {
  "func"() ( {
  ^bb0(%arg0: memref<4x64x128x3xf32>, %arg1: memref<5x3x3x10xf32>):  // no predecessors
    %cst = "std.constant"() {name = "constant.5", value = dense<0.000000e+00> : tensor<f32>} : () -> tensor<f32>
    %0 = "std.alloc"() : () -> memref<4x30x42x10xf32>
    "xla_lhlo.broadcast_in_dim"(%cst, %0) {broadcast_dimensions = dense<[]> : tensor<0xi64>, name = "broadcast.6"} : (tensor<f32>, memref<4x30x42x10xf32>) -> ()
    %1 = "std.alloc"() : () -> memref<4x30x42x10xf32>
    "xla_lhlo.convolution"(%arg0, %arg1, %1) {batch_group_count = 1 : i64, dimension_numbers = {input_batch_dimension = 0 : i64, input_feature_dimension = 3 : i64, input_spatial_dimensions = dense<[1, 2]> : tensor<2xi64>, kernel_input_feature_dimension = 2 : i64, kernel_output_feature_dimension = 3 : i64, kernel_spatial_dimensions = dense<[0, 1]> : tensor<2xi64>, output_batch_dimension = 0 : i64, output_feature_dimension = 3 : i64, output_spatial_dimensions = dense<[1, 2]> : tensor<2xi64>}, feature_group_count = 1 : i64, lhs_dilations = dense<1> : tensor<2xi64>, name = "convolution.4", padding = dense<0> : tensor<2x2xi64>, precision_config = ["DEFAULT", "DEFAULT"], rhs_dilations = dense<1> : tensor<2xi64>, window_strides = dense<[2, 3]> : tensor<2xi64>} : (memref<4x64x128x3xf32>, memref<5x3x3x10xf32>, memref<4x30x42x10xf32>) -> ()
    %2 = "std.alloc"() : () -> memref<4x30x42x10xf32>
    "xla_lhlo.maximum"(%0, %1, %2) {name = "maximum.7"} : (memref<4x30x42x10xf32>, memref<4x30x42x10xf32>, memref<4x30x42x10xf32>) -> ()
    "xla_lhlo.terminator"() : () -> ()
  }) {sym_name = "main", type = (memref<4x64x128x3xf32>, memref<5x3x3x10xf32>) -> ()} : () -> ()
  "module_terminator"() : () -> ()
}) : () -> ()
```
This is nothing specific to broadcast_in_dim (happens with say xla_hlo.add as well). This is likely a missing conversion for std.constant that needs to be completed?

### PR DESCRIPTION
**Pull request by bondhugula**
Add a conversion pattern to convert std.constant to xla_lhlo.const - to
run during HLO to LHLO conversion. Without such a pattern, xla_hlo to
xla_lhlo conversion would fail when std.constant ops provide constant tensor
operands to other xla_hlo ops. The tf-mlir-translate -hlo-to-mlir-hlo
for example generates such std.constant ops from HLO constant nodes.

std.constant tensor generating ops are to be replaced by constant memref
generating ops.

Fixes #39895.

Signed-off-by: Uday Bondhugula <uday@polymagelabs.com>

### CODE REVIEW

**Comment by bondhugula** [comment]
@MaheshRavishankar @sherhut @pifon2a for visibility.

File: tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir [hunk 1]
```diff
@@ -432,7 +432,23 @@ func @dot(%arg0: tensor<1024x1024xf32>) -> tensor<1024x1024xf32> {
...
+  %dot = "xla_hlo.dot"(%arg0, %arg0)
+     : (tensor<1024x1024xf32>, tensor<1024x1024xf32>) -> tensor<1024x1024xf32>
+  return %dot : tensor<1024x1024xf32>
+}
+
+// -----
+
+// Test whether std.constant's generating tensors are converted to lhlo.const.
+func @std.const.tensor() {
+  %cst = constant  dense<0.000000e+00> : tensor<f32>
```
Comment (pifon2a) [inline tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir:444]: why is it not an `xla_hlo.constant`?

Comment (bondhugula) [reply tensorflow/compiler/mlir/xla/tests/hlo-legalize-to-lhlo.mlir:444]: > why is it not an `xla_hlo.constant`?

Please see here: https://github.com/tensorflow/tensorflow/issues/39895

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 2]
```diff
@@ -310,6 +311,32 @@ class HloToLhloTensorStoreOpConverter
...
+  LogicalResult matchAndRewrite(
+      mlir::ConstantOp constOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter& rewriter) const final {
+    Operation* op = constOp.getOperation();
+    OpResult result = op->getResult(0);
+    RankedTensorType resultType = result.getType().dyn_cast<RankedTensorType>();
+    if (!resultType) return failure();
+
+    // Tensors generated by constant ops are always statically shaped.
+    CHECK(resultType.hasStaticShape());
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc:329]: Seems like an assert() is more suitable in MLIR code.

Comment (bondhugula) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc:329]: Done.

**Comment by joker-eph** [comment]
(discussion is going on in the tracking bug  #39895)

**Comment by gbaned** [comment]
@bondhugula, @joker-eph This PR is in draft, any update on this? Please. Thanks!

**Comment by gbaned** [comment]
@bondhugula, @joker-eph This PR is in draft, any update on this? Please. Thanks!

**Comment by bondhugula** [comment]
I think this PR can be closed in favor of having a pattern to convert 

> @bondhugula, @joker-eph This PR is in draft, any update on this? Please. Thanks!

I just posted on the other issue thread to resurrect discussion on  #39895

**Comment by gbaned** [comment]
> > @bondhugula, @joker-eph This PR is in draft, any update on this? Please. Thanks!
> 
> I just posted on the other issue thread #39895 to resurrect discussion.

@bondhugula  Any update on this PR? Please. Thanks!

**Comment by bondhugula** [comment]
There isn't a clear solution to this issue and this has become moot / low priority - given that there are ways to avoid it trivially. I'm closing this. The bug it was trying to fix should remain open however - the behavior / error isn't proper.

**Pull request closed without merging**